### PR TITLE
Async local database

### PR DESCRIPTION
**Pull request by FirelightFlagboy**
Supersede #3848

### CODE REVIEW

File: oxidation/libparsec/crates/core_fs/src/storage/chunk_storage.rs [hunk 1]
```diff
@@ -34,72 +35,81 @@ struct NewChunk<'a> {
...
+            )
+            .await
     }
 
-    fn get_local_chunk_ids(&self, chunk_ids: &[ChunkID]) -> FSResult<Vec<ChunkID>> {
+    async fn get_local_chunk_ids(&self, chunk_ids: &[ChunkID]) -> FSResult<Vec<ChunkID>> {
         let bytes_id_list = chunk_ids
             .iter()
-            .map(|chunk_id| (**chunk_id).as_ref())
+            .map(|chunk_id| (**chunk_id).as_ref().to_vec())
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/core_fs/src/storage/chunk_storage.rs:112]: Who do you need to allocate ?

File: oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs [hunk 2]
```diff
@@ -460,21 +495,22 @@ impl ManifestStorage {
...
         };
-        cache_entry.and_then(|entry| {
-            let unlock = entry.lock().expect("Mutex is poisoned");
-            unlock.manifest.clone()
-        })
+        if let Some(entry) = cache_entry {
+            block_on(entry.lock()).manifest.clone()
+        } else {
+            None
+        }
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:502]:
```rust
cache_entry.map(|x| block_on(x.lock()).manifest.clone())
```

File: oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs [hunk 3]
```diff
@@ -485,56 +521,58 @@ impl ManifestStorage {
...
-            (&mut cache.manifest, &mut cache.pending_chunk_ids)
+        if let (Some(manifest), Some(pending_chunk_ids)) =
+            (cache.manifest.clone(), cache.pending_chunk_ids.clone())
         {
             let ciphered = manifest.dump_and_encrypt(&self.local_symkey);
-            let pending_chunk_ids = pending_chunk_ids
+            let chunk_ids = pending_chunk_ids
                 .iter()
-                .map(|chunk_id| chunk_id.as_ref())
+                .map(|chunk_id| chunk_id.as_ref().to_vec())
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:566]: Why allocate ?

File: oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs [hunk 4]
```diff
@@ -548,64 +586,80 @@ impl ManifestStorage {
...
+            .conn
+            .exec(move |conn| {
+                let vlob_id = (*entry_id).as_ref();
+                diesel::delete(vlobs::table.filter(vlobs::vlob_id.eq(vlob_id))).execute(conn)
+            })
+            .await?
+            > 0;
 
-        if let Some(pending_chunk_ids) = &cache_unlock.pending_chunk_ids {
+        if let Some(pending_chunk_ids) = cache_unlock.pending_chunk_ids.clone() {
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:645]: Why clone ?

File: oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs [hunk 5]
```diff
@@ -548,64 +586,80 @@ impl ManifestStorage {
...
             let pending_chunk_ids = pending_chunk_ids
                 .iter()
-                .map(ChunkOrBlockID::as_ref)
+                .map(|chunk| chunk.as_ref().to_vec())
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:648]: Why allocate ?

File: oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs (hunk 3 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:550]: The value is used when we delete the chunks from the table.
And rust require that the value live for long enough

File: oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs (hunk 2 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs]: This method no longer exist ;)

File: oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs (hunk 4 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:629]: Lifetime concern when sending the value across thread

File: oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs (hunk 3 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:550]: That the requirement for sending value across thread, it require the `'static` lifetime :/

File: oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs (hunk 4 above)
Comment (TimeEngineer) [reply oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:629]: Which thread ?

Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:629]: take a look to `platform_local_db`

File: oxidation/libparsec/crates/platform_local_db/src/executor.rs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+use std::thread::JoinHandle;
+
+use diesel::SqliteConnection;
+use platform_async::channel;
+
+use crate::{DatabaseError, DatabaseResult};
+
+pub struct SqliteExecutor {
+    job_sender: Option<channel::Sender<ExecOrStop>>,
+    handle: Option<JoinHandle<()>>,
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/platform_local_db/src/executor.rs:12]: Why do you need `Option` for both ?

Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/platform_local_db/src/executor.rs:12]: The option on `job_sender` isn't required, I forgot to remove it.
The option on `handle` is required to be able to `join` it in the `Drop` (the function `join` take the ownership of the value)

Comment (TimeEngineer) [reply oxidation/libparsec/crates/platform_local_db/src/executor.rs:12]: You can `join` without `Option`

File: oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs [hunk 7]
```diff
@@ -548,64 +570,80 @@ impl ManifestStorage {
...
+            }).await?;
+
+            for pending_chunk_id in chunk_ids
+                .chunks(LOCAL_DATABASE_MAX_VARIABLE_NUMBER)
+                .map(|chunk| chunk.to_vec())
+            {
+                self.conn
+                    .exec(move |conn| {
+                        diesel::delete(
+                            chunks::table.filter(chunks::chunk_id.eq_any(pending_chunk_id)),
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:583]: use `pending_chunk_id.iter().map(ChunkOrBlockID::as_ref)` there, which replace the `map` and avoid the `.to_vec()`

File: oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs [hunk 8]
```diff
@@ -548,64 +570,80 @@ impl ManifestStorage {
...
+            })
+            .await?
+            > 0;
 
-        if let Some(pending_chunk_ids) = &cache_unlock.pending_chunk_ids {
+        if let Some(pending_chunk_ids) = cache_unlock.pending_chunk_ids.clone() {
             let pending_chunk_ids = pending_chunk_ids
                 .iter()
-                .map(ChunkOrBlockID::as_ref)
+                .map(|chunk| chunk.as_ref().to_vec())
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:632]: Same `map` once you are in `exec` and use `into_iter()`

File: oxidation/libparsec/crates/core_fs/src/storage/workspace_storage.rs [hunk 9]
```diff
@@ -455,13 +503,18 @@ impl WorkspaceStorageSnapshot {
...
-                _ => Err(FSError::LocalMiss(*entry_id)),
-            },
-            None => Err(FSError::InvalidFileDescriptor(fd)),
+    pub async fn load_file_descriptor(&self, fd: FileDescriptor) -> FSResult<LocalFileManifest> {
+        let entry_id = self
+            .open_fds
+            .lock()
+            .expect("Mutex is poisoned")
+            .get(&fd)
+            .cloned()
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/core_fs/src/storage/workspace_storage.rs:512]: `.copied()` instead ?

File: oxidation/libparsec/crates/platform_local_db/tests/book/mod.rs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+pub async fn create_table(local_db: &LocalDatabase) {
+    let res = local_db
+        .exec(|conn| {
+            sql_query(
+                "
+                    CREATE TABLE books (
+                        id INTEGER PRIMARY KEY NOT NULL,
+                        name TEXT NOT NULL,
+                        liked BOOLEAN
+                    )
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/platform_local_db/tests/book/mod.rs:30]: Indentation

File: oxidation/libparsec/crates/core_fs/src/storage/workspace_storage.rs [hunk 11]
```diff
@@ -531,29 +581,33 @@ impl WorkspaceStorageSnapshot {
...
 
-    pub fn lock_entry_id(&self, entry_id: EntryID) -> Arc<Mutex<()>> {
-        self.workspace_storage.lock_entry_id(entry_id)
+    pub async fn lock_entry_id(&self, entry_id: EntryID) -> Arc<AsyncMutex<()>> {
+        self.workspace_storage.lock_entry_id(entry_id).await
     }
 
-    pub fn release_entry_id(&self, entry_id: EntryID, guard: MutexGuard<()>) {
-        self.workspace_storage.release_entry_id(entry_id, guard)
+    pub async fn release_entry_id<'a>(&self, entry_id: EntryID, guard: AsyncMutexGuard<'a, ()>) {
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/core_fs/src/storage/workspace_storage.rs:602]: lifetime

**Review by TimeEngineer — CHANGES_REQUESTED** [review]

File: oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs [hunk 12]
```diff
@@ -485,56 +505,58 @@ impl ManifestStorage {
...
-        self.ensure_manifest_persistent_unlock(
-            entry_id,
-            &mut cache_lock.lock().expect("Mutex is poisoned"),
-        )
+        self.ensure_manifest_persistent_unlock(entry_id, cache_lock.lock().await)
+            .await
     }
 
-    fn ensure_manifest_persistent_unlock(
+    async fn ensure_manifest_persistent_unlock<'a>(
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:539]: It's used for `mut cache`

File: oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs [hunk 13]
```diff
@@ -485,56 +505,58 @@ impl ManifestStorage {
...
         &self,
         entry_id: EntryID,
-        cache: &mut CacheEntry,
+        mut cache: AsyncMutexGuard<'a, CacheEntry>,
     ) -> FSResult<()> {
-        if let (Some(manifest), Some(ref pending_chunk_ids)) =
-            (&mut cache.manifest, &mut cache.pending_chunk_ids)
+        if let (Some(manifest), Some(pending_chunk_ids)) =
+            (cache.manifest.clone(), cache.pending_chunk_ids.clone())
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:545]: For that I need to extract `need_sync` and `base_version` that seems doable :+1:

File: oxidation/libparsec/crates/core_fs/src/storage/workspace_storage.rs [hunk 14]
```diff
@@ -34,26 +35,29 @@ lazy_static::lazy_static! {
...
-        map.entry(id).or_insert_with(|| Arc::new(Mutex::new(())));
+    async fn acquire(&self, id: T) -> Arc<AsyncMutex<()>> {
+        let mut map = self.0.lock().await;
+        map.entry(id)
+            .or_insert_with(|| Arc::new(AsyncMutex::new(())));
         map.get(&id).unwrap_or_else(|| unreachable!()).clone()
     }
-    fn release(&self, id: T, guard: MutexGuard<()>) {
+
+    async fn release<'a>(&self, id: T, guard: AsyncMutexGuard<'a, ()>) {
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/core_fs/src/storage/workspace_storage.rs:53]: Used by `guard`

File: oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:539]: `mut cache: AsyncMutexGuard<'_, CacheEntry>`

File: oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs (hunk 8 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:632]: done

**Review by TimeEngineer — APPROVED** [review]

**Pull request merged**